ariadne/cardano/src/Ariadne/MainTemplate.hs

```diff
@@ -153,12 +155,13 @@ initializeEverything MainSettings {..}
 
     serviceAction :: IO ()
     serviceAction =
-      raceWithUpdateCheckAction $
-      cardanoAction
+      raceWithUpdateCheckAction $ cardanoAction
 
     mainAction :: IO ()
     mainAction = do
       initAction
+      -- check if there are unknown keys in keyfile.
+      async walletCheckKeys
```

Comment:
`async` doesn't kill the thread it spawns when its parent thread dies. It's generally safer to use `withAsync` unless there is a specific reason to use `async`.
See https://www.notion.so/serokell/Concurrency-f0ad92790b5c4478a8f22bf5b4716fc8

Reply:
Ok, thanks. Wanted to ask about that, but forget it.

Reply:
can't run `walletCheckKeys` in `walletInitAction` using `withAsync` - walletCheckKeys waits for confirmation from UI, but when inner action returns, thread with  `walletCheckKeys` will be killed before getting confirmation from UI. 

So - I can run it with async, or run the whole `walletInitAction` asynchronously, using `withAsync`. 
I prefer the last variant.

Reply:
Hm, and now I tend to think that `walletCheckKeys` is closer to service action than init action, because it should be launched after everything (including UI) is initialized, not before.
So probably your first version (returning three things from `wpMakeWallet`) was better…
But let's not call it `walletCheckKeys` here, let's call it `walletAction`, `defaultMain` doesn't need to know what exactly it does, it only needs to know that it's some action that needs to be run after initialization.

Reply:
@Haskell-mouse have you seen it? ↑

Reply:
Yes, sorry, done smth similar in another PR.